### DISCUSSION TITLE
Library Folder Structure

### DISCUSSION THREAD

**Discussion by Puddl36lum** [openaudible/openaudible#800] [Ideas]
Is there a way create folder structure within the library folder and still have openaudible sync?  I have a very large audible library and i'd like a little more organization.  I usually organize by author so my organization would look like:

OpenAudible>books>Jim Butcher>Codex Alera> audiobooks.

**Comment by christyler80** [comment]
Author > Title folder structure would also be good for those of us who use Plex.

**Comment by pgx3d** [comment]
Hi

I would really appreciate getting folder structure options.

This is the reason I'm not using OpenAudible anymore.

**Reply by lotherk** [reply]
You have a better alternative than OpenAudible?

**Comment by openaudible** [comment]
We are definitely working on this. We don't want to break anything for existing users.. so its a slow and careful process. 

There are a lot of edge cases that can make this tricky (and not great for the user) -- multiple authors, illegal filename characters, series with mismatched names, case sensitivity issues, and duplicate filename destinations, to name a few. 

The best thing to do is have a good audiobook player that lets you sort your library the way you want or use the search. 

We know this is a popular feature request and will keep working on it. Have been trying to stay focused making the core feature set rock solid, fast and user-friendly.

**Comment by christyler80** [comment]
Hi,

I have built an Excel macro that scans the folder with the mp3 files created using OpenAudible and moves (or copies) these files into an "Author > Book Title" folder structure. 

The code is fairly straightforward, using Microsoft's "Scripting.FileSystemObject". If you are interested, I will 'tidy it up' and upload it to GitHub in the next few weeks.

Regards

Chris

**Comment by mediacowboy** [comment]
I would love for this to be able to automatically do folder structure.

For Series:
Author > Series > Book {Series#} - Title > (automatic split chapters)

Non-Series:
Author > Title > (automatically split chapters)

**Comment by lotherk** [comment]
Any updates on this?

**Comment by openaudible** [comment]
An update.... 

We've done lots of internal re-writing and testing to possibly support it in the future, but its a minefield of edge cases that can make this feature a disaster. 
Problems: 
- Clutters up the UI. 
- Gives the user too many choices and ability to make bad choices 
- Makes it hard to change the structure (for instance, if a file is open in a player and can't be moved.) 
- Makes it harder to find your books if the tags are not perfect (author can be slightly different, or contain multiple authors.)

Recommend: 
- Use the app and right click, show MP3/M4B
- Use a good file system file finder (Everything for PC)
- Use a web front end like "AudioBookShelf". 

The latest beta has a number of new features that we are working on getting released. That includes better sorting for authors, series, and series number. Improved searching. Export to Excel. Better M4B tagging. Better file import. 
https://openaudible.org/beta

I am hoping the beta adds tags that can be read by Plex and another players... but we might need to look at the source code for plex and recommend a change to show the newest info.

**Reply by lotherk** [reply]
Thank you for your reply. 

I don't think what we want clutters up the UI. It is basically an input field where we can define a format like %A/%N/%A-%N.m4br which would be Artist/Name/Artist - Name.m4b or something. I also doubt it gives us too many choices and abilities to make bad choices (not sure what you think about your users here)... 
It does not make it hard to change structure and a file open in a player because why would it? It is for new books to be downloaded and written after converting. I also don't think it makes it harder to find my book, at least not harder than it is now.

Your recommandations don't help at all with what we want here, or maybe I misunderstood what the others want in the first place. 

Again, it is not about finding files, it is about having a folder structure that other software can use, for example Plex. Like this, if I add my "books" folder to plex, it appears as one audiobook with 63 chapters (I have 63 audiobooks). If I re-organize my audiobooks in folders, manually, OpenAudible wants to download and convert them again (because I moved the files and it can not find it anymore). Of course, I can copy it and organize it, which, currently, wastes 50GB of my precious space.

**Reply by openaudible** [reply]
Custom filenames are still on the feature list. We might make some kind of "One book per folder" checkbox and if the user selects that, we can allow a file format. But just trying to update why it is a pain in the butt and we're not super thrilled about causing ourselves a lot of headaches. 

As far as plex goes, maybe a plugin or agent would be a better way to import audiobooks into plex. (But, correct me if I'm wrong, but looking online, it looks like plex extensions are not well supported.) 
This repo [github.com/macr0dev/Audiobooks.bu](https://github.com/macr0dev/Audiobooks.bundle) could be modified to import files from our existing books.json file. Importing from JSON would be a lot easier than trying to scrape the data from audible, which is what the plugin above attempts. 

We also store a lot of metadata about each book in the audiobook files. Some kind of media importer could also be used. Will look at that source code and maybe we can find a python plex plugin programmer. (Anyone??)

**Comment by christyler80** [comment]
Hi,

I have an Excel spreadsheet that scans the OpenAudible download folder and makes a copy of the MP3 files into a different location using Author->Title folder structure.

This way, the OpenAudible side of things is untouched, although it is a bit wasteful on disk space having two copies of every book. 

There is an option in my code to use symbolic links (which I use), but this going wrong would impact OpenAudible downloads, so I am not inclined to enable this by default.

I will upload the spreadsheet to my GitHub pages in the next couple of weeks, after I have written some basic documentation and let you know when this is available.

The code is written in VBA and not particularly complicated. Happy to speak with the OpenAudible author(s) about converting the code to a different language if that adds value.

Best Regards

Chris

**Comment by christyler80** [comment]
Hi,

 

Excel Media Manager can be downloaded from https://github.com/christyler80/ExcelMediaManager.

 

For Open Audible compatibility, the code supports moving files from a specified ‘Source / Downloads’ folder, into a ‘Library’ folder structure in a different location.

*        When this move option is taken, the original file in the ‘Source / Downloads’ folder is replaced by a  <https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/mklink> ‘Symbolic Link’ to the new file location under the ‘Library’ folder structure.

*        This allows Open Audible to recognise files that have already been downloaded, also media managers such as Plex, access to the MP3 (et al) file properties.

 

Best Regards

 

Chris

 

PS, Please make sure that you have backups of your existing media files and folders before trying this program.

 

From: ***@***.*** ***@***.***> 
Sent: Tuesday, March 7, 2023 5:09 PM
To: 'openaudible/openaudible' ***@***.***>; 'Konrad Lother' ***@***.***>
Subject: RE: [openaudible/openaudible] Library Folder Structure (Discussion #800)

 

Hi,

 

I have an Excel spreadsheet that scans the OpenAudible download folder and makes a copy of the MP3 files into a different location using Author->Title folder structure.

 

This way, the OpenAudible side of things is untouched, although it is a bit wasteful on disk space having two copies of every book. 

 

There is an option in my code to use symbolic links (which I use), but this going wrong would impact OpenAudible downloads, so I am not inclined to enable this by default.

 

I will upload the spreadsheet to my GitHub pages in the next couple of weeks, after I have written some basic documentation and let you know when this is available.

 

The code is written in VBA and not particularly complicated. Happy to speak with the OpenAudible author(s) about converting the code to a different language if that adds value.

 

Best Regards

 

Chris

**Comment by orbitalteapot** [comment]
I don't know if anyone needs this, but i hope this can help somebody.

https://github.com/orbitalteapot/OpenAudibleBookManager

**Comment by christyler80** [comment]
Hi,

 

Thank you and well done on a nice bit of coding.

 

I built something equivalent using Microsoft Excel (ExcelMediaManager <https://github.com/christyler80/ExcelMediaManager> ) and found that I needed to leave a Symbolic Link ‘symlink‘ file in the OpenAudible downloads folder pointing to the ‘new’ file location, for OpenAudible to recognise that the book had already been downloaded.

 

Not sure if this is relevant to your solution, but thought to mention it.

 

Regards

 

Chris

**Comment by MoSattler** [comment]
Smal python script to organise the mb4 files, maybe it's helpful to someone

```python
import os
from mutagen.mp4 import MP4

def organize_m4b(source_directory):
    for root, dirs, files in os.walk(source_directory):
        for file in files:
            if file.endswith(".m4b"):
                file_path = os.path.join(root, file)
                metadata = MP4(file_path).tags
                # Example: Use '©ART' for artist/author, '©alb' for album/book title
                author = metadata.get('©ART', ['Unknown Author'])[0]
                book_title = metadata.get('©alb', ['Unknown Title'])[0]
                # Define new path
                new_dir = os.path.join(source_directory, author, book_title)
                os.makedirs(new_dir, exist_ok=True)
                new_file_path = os.path.join(new_dir, file)
                os.rename(file_path, new_file_path)
                print(f'Moved "{file}" to "{new_dir}"')

source_dir = '/path/to/your/m4b/files'
organize_m4b(source_dir)
```

**Comment by skymakai** [comment]
I'm surprised this issue is still open 4+ years later.  I came looking for a way to use both Plex and OpenAudible.  Plex simply needs audiobooks to be in Author Folders.  It seems to me that OpenAudible simply needs to make Author folders and it's database to know to look for files within folders.

**Reply by MoSattler** [reply]
same use case here
